### ISSUE TITLE
rsyslogd fails to start because of segmentation fault

### ISSUE THREAD

**Issue by Rampa33** [rsyslog/rsyslog#3859]
### Expected behavior
I just added a configuration for main_queue
main_queue(queue.filename="myqueue" queue.saveOnShutdown="on")
to my rsyslog.conf. Without this line, rsyslogd is starting and forwarding logs to my remote machines.

### Actual behavior
With the above config, i am seeing a segmentation fault.

### Environment
-version : 
~ # rsyslogd -v
rsyslogd  8.1905.0 (aka 2019.05) compiled with:
	PLATFORM:				x86_64-openwrt-linux-gnu
	PLATFORM (lsb_release -d):		Description:	Ubuntu 14.04 LTS
	FEATURE_REGEXP:				Yes
	GSSAPI Kerberos 5 support:		No
	FEATURE_DEBUG (debug build, slow code):	No
	32bit Atomic operations supported:	No
	64bit Atomic operations supported:	No
	memory allocator:			system default
	Runtime Instrumentation (slow code):	No
	uuid support:				Yes
	systemd support:			No
	Config file:				/etc/rsyslog.conf
	PID file:				/var/run/rsyslogd.pid
	Number of Bits in RainerScript integers: 64

**Comment by Rampa33** [comment]
4452.171080908:main thread    : ../action.c: Action builtin:omfwd[0xf43840]: queue 0xf43f00 started
4452.171085339:main thread    : ruleset.c: Activating Ruleset Queue[(nil)] for Ruleset RSYSLOG_DefaultRuleset
4452.171093193:main thread    : rsconf.c: activateMainQueue: mainq cnf obj ptr is 0xf38c10
4452.171099758:main thread    : rainerscript.c: nvlstGetParam: name 'queue.filename', type 13, valnode->bUsed 0
4452.171137670:main thread    : rainerscript.c: nvlstGetParam: name 'queue.saveonshutdown', type 4, valnode->bUsed 0
4452.171148562:main thread    : queue.c: queue param blk:
4452.171154381:main thread    : rainerscript.c: queue.filename:  'myqueue'
4452.171162195:main thread    : rainerscript.c: queue.spooldirectory: (unset)
4452.171168213:main thread    : rainerscript.c: queue.size: (unset)
4452.171177803:main thread    : rainerscript.c: queue.dequeuebatchsize: (unset)
4452.171190196:main thread    : rainerscript.c: queue.mindequeuebatchsize: (unset)
4452.171212668:main thread    : rainerscript.c: queue.mindequeuebatchsize.timeout: (unset)
4452.171220978:main thread    : rainerscript.c: queue.maxdiskspace: (unset)
4452.171227841:main thread    : rainerscript.c: queue.highwatermark: (unset)
4452.171236278:main thread    : rainerscript.c: queue.lowwatermark: (unset)
4452.171242270:main thread    : rainerscript.c: queue.fulldelaymark: (unset)
4452.171248360:main thread    : rainerscript.c: queue.lightdelaymark: (unset)
4452.171254457:main thread    : rainerscript.c: queue.discardmark: (unset)
4452.171292588:main thread    : rainerscript.c: queue.discardseverity: (unset)
4452.171302356:main thread    : rainerscript.c: queue.checkpointinterval: (unset)
4452.171310698:main thread    : rainerscript.c: queue.syncqueuefiles: (unset)
4452.171316997:main thread    : rainerscript.c: queue.type: (unset)
4452.171322843:main thread    : rainerscript.c: queue.workerthreads: (unset)
4452.171332569:main thread    : rainerscript.c: queue.timeoutshutdown: (unset)
4452.171342080:main thread    : rainerscript.c: queue.timeoutactioncompletion: (unset)
4452.171363307:main thread    : rainerscript.c: queue.timeoutenqueue: (unset)
4452.171376955:main thread    : rainerscript.c: queue.timeoutworkerthreadshutdown: (unset)
4452.171388456:main thread    : rainerscript.c: queue.workerthreadminimummessages: (unset)
4452.171394996:main thread    : rainerscript.c: queue.maxfilesize: (unset)
4452.171401108:main thread    : rainerscript.c: queue.saveonshutdown: 1
4452.171411105:main thread    : rainerscript.c: queue.dequeueslowdown: (unset)
4452.171450057:main thread    : rainerscript.c: queue.dequeuetimebegin: (unset)
4452.171456043:main thread    : rainerscript.c: queue.dequeuetimeend: (unset)
4452.171465268:main thread    : rainerscript.c: queue.cry.provider: (unset)
4452.171471803:main thread    : rainerscript.c: queue.samplinginterval: (unset)
Segmentation fault
~ #


This is the log i see when i try to do 'rsyslogd -dn'. Why other fields of the main queue are not taking the default value?

notify @rgerhards @davidelang

**Comment by davidelang** [comment]
I think you need to set the workdirectory (or spool directory) for it to ork 
properly.

David Lang

**Comment by Rampa33** [comment]
@davidelang ,

Thanks for your response. My main_queue() configuration is infact preceded by global(workDirectory="/path/to/directory") statement.

i also do have action specific queues, which are picking up workdirectory config.

**Comment by davidelang** [comment]
I would put all the global() parameters in one set. I don't know how rsyslog 
handles having multiple global() statements.

David Lang

**Comment by rgerhards** [comment]
El sáb., 7 sept. 2019 a las 9:53, David Lang
(<notifications@github.com>) escribió:
>
> I would put all the global() parameters in one set. I don't know how rsyslog
> handles having multiple global() statements.

That's well defined and OK - we know we need to have them split from
time to time (especially when using includes).

I'd still like to see the full config - sorry, right now to busy to
try to reproduce albeit I know it's trivial to do...

Rainer

**Comment by Rampa33** [comment]
Here is my complete config file:


module(load="imuxsock")
module(load="imklog")
module(load="impstats" interval="10" severity="7")

$ActionFileDefaultTemplate RSYSLOG_DebugFormat 

if $syslogtag contains 'rsyslogd-pstats' then {
     action(type="omfile" queue.type="linkedlist" queue.discardmark="980"
            name="pstats" file="/var/log/pstats")
     stop
}

*.*     /var/log/messages
global(workDirectory="/var/rampa/")
main_queue(queue.filename="myq" queue.saveOnShutdown="on")

# Rules for forwarding a specific subset to a remote server go here.

local3.DEBUG action(type="omfwd" target="10.100.3.100" port="514" address="169.254.9.1" protocol="TCP" queue.type="Disk" queue.filename="rampa" action.resumeRetryCount="-1" queue.size="100" queue.timeoutEnqueue="0" queue.checkpointInterval="1" queue.syncqueuefiles="on")

local3.DEBUG action(type="omfwd" target="10.200.4.200" port="514" address="169.254.9.1" protocol="TCP")

**Comment by rgerhards** [comment]
Thanks for the report, and apologies for the very late follow-up.

This is old enough that the result may no longer be useful for the original OpenWrt build, but I am adding the current finding for people who arrive here via search while we work through the backlog. We do not lightly close issues and do not close them just because of age.

I checked the reported main-queue shape against current `main` (`8.2606.0.daily`). The current queue construction path keeps the configured global `workDirectory` as the queue spool directory unless `queue.spoolDirectory` is explicitly set, then applies the `main_queue(...)` parameters before starting the queue.

Validation performed on current `main`:

- `main_queue(queue.filename="myq" queue.saveOnShutdown="on")` with a configured `global(workDirectory=...)` passes `rsyslogd -N1`.
- A closer config shape with the reported `main_queue(...)` plus two `omfwd` actions, one using a disk queue, also passes `rsyslogd -N1`.
- Starting that config in foreground mode stayed alive until an external `timeout` sent SIGTERM; it did not segfault during startup.

So I am closing this as fixed in current rsyslog. If a current release still crashes with this configuration, please open a new issue with the current version, full config, platform/toolchain details, and a fresh backtrace or sanitizer/Valgrind output.
